### ISSUE TITLE
Non-namespaced macro pollutes the global namespace in a public header

### ISSUE THREAD

**Issue by mhoemmen** [tlapack/tlapack#6]
https://github.com/tlapack/tlapack/blob/9887d9ad9251b5549290a3c163efca21facbb057/include/lapack/geqr2.hpp#L35

The lapack/geqr2.hpp header unconditionally defines a macro `A(i_, j_)`.  Even though the macro is undefined on line 66 in the same header, this is a very short name that application could would be likely to define.  If an application defines that macro and includes this header, a compiler error would result.

Given that this is a C++ project, not a C project, would you consider either using [mdspan](https://github.com/kokkos/mdspan/) to 
represent multidimensional arrays, or writing your own matrix view class for internal use?

**Comment by weslleyspereira** [comment]
Hi @mhoemmen. Thank you very much for the comments!

I agree with you that it is better to protect the macros. I took a look at the mdspan. Can we use it with the C++11 standard? We didn't plan to have our own matrix view class. Thanks again.

**Comment by luszczek** [comment]
I think I heard my name, thank you @mhoemmen  I suggest to create a compiler compatibility table for [mdspan](https://github.com/kokkos/mdspan/) to assuage any reservations. I think most vendor compilers are on board with C++14 now. And let's hope `<mdspan>` becomes part of C++ standard library. Are Cray and IBM compilers supported?

**Comment by mhoemmen** [comment]
Hi @luszczek !  Always great to hear from you! : - )

Compiler requirements for the reference implementation of `mdspan` are tied somewhat to Kokkos' requirements.  Kokkos has a "shifting language window" model to minimize testing burden.  While I'm not a Kokkos developer (any more), I imagine that it wants to support the same compilers that you do.

There are lovely tables of compilers and supported language features [on the cppreference website](https://en.cppreference.com/w/cpp/compiler_support).

If you really must support C++ versions less than 14, I would recommend submitting a patch to the mdspan reference implementation.  It's so much more productive to have actual multidimensional arrays, and you'll be able to take row-major or column-major (or strided, or Z-ordered, or ...) arrays with a single implementation.

**Comment by weslleyspereira** [comment]
Thanks for the table! It is good to have it in mind. So, I have no problems moving to the C++14 standard.
Also, I will try the mdspan and give some feedback here. I agree it is a good way to go. It is good to have things well designed at the beginning of a new project.

Thanks!

**Comment by mhoemmen** [comment]
@weslleyspereira wrote:
> Thanks!

Gladly!  The [reference implementation](https://github.com/kokkos/stdBLAS) of our C++ BLAS proposal relies on C++20, but you might like to imitate its interface.  I would also consider pull requests to make the implementation compatible with C++14 (e.g., by replacing direct use of `requires` constraints with a macro that dispatches to `std::enable_if_t` if constraints are not available).

**Comment by weslleyspereira** [comment]
> Gladly! The [reference implementation](https://github.com/kokkos/stdBLAS) of our C++ BLAS proposal relies on C++20, but you might like to imitate its interface. I would also consider pull requests to make the implementation compatible with C++14 (e.g., by replacing direct use of `requires` constraints with a macro that dispatches to `std::enable_if_t` if constraints are not available).

I will certainly take a look :)

**Comment by weslleyspereira** [comment]
Hi @mhoemmen. This issue was the first step that led me to the changes in the interface I implemented in #7. Thank you and @luszczek for the discussion!!

To summarize: After many rounds of discussions, we opted for an abstract interfaces like:

```c++
template< class uplo_t, class matrix_t,
    enable_if_t<(
    /* Requires: */
        is_same_v< uplo_t, upper_triangle_t > || 
        is_same_v< uplo_t, lower_triangle_t >
    ), int > = 0
>
int potrf2( uplo_t uplo, matrix_t& A ) { ... }
```

This was strongly based on [P1673](https://wg21.link/p1673), but I also took some ideas from kokkos/mdspan.

Since the main concern of #6 was solved by #7, I will merge it to the master branch, and this issue will be closed. I will try to give more details about the new interface in the documentation in the near future.
